Sources/DataPipeline/DataPipelineImplementation.swift

```diff
@@ -68,31 +87,77 @@ class DataPipelineImplementation: DataPipelineInstance {
             return attributesPlugin?.attributes ?? [:]
         }
         set {
-            if let attributesPlugin = analytics.find(pluginType: DeviceAttributes.self) {
-                attributesPlugin.attributes = newValue
-            } else {
-                // TODO: [CDP] Verify with server's expectation
-                let attributesPlugin = DeviceAttributes()
-                attributesPlugin.attributes = newValue
-                analytics.add(plugin: attributesPlugin)
-            }
+            addDeviceAttributes(newValue)
+        }
+    }
+
+    /// Adds device default and custom attributes using DeviceAttributes plugin
+    private func addDeviceAttributes(_ customAttributes: [String: Any] = [:]) {
+        // Consolidate all Apple platforms under iOS
+        let deviceOsName = "iOS"
+        // FIXME: [CDP] Fetch the right defaultDeviceAttributes here
+        // deviceAttributesProvider.getDefaultDeviceAttributes { defaultDeviceAttributes in
+        let defaultDeviceAttributes: [String: Any] = [:]
+        let deviceAttributes: [String: Any] = defaultDeviceAttributes
+            .mergeWith([
+                "platform": deviceOsName,
+                "lastUsed": dateUtil.now
+            ])
+            .mergeWith(customAttributes)
+
+        if let attributesPlugin = analytics.find(pluginType: DeviceAttributes.self) {
+            attributesPlugin.attributes = deviceAttributes
+        } else {
+            // TODO: [CDP] Verify with server's expectation
+            let attributesPlugin = DeviceAttributes()
+            attributesPlugin.attributes = deviceAttributes
+            analytics.add(plugin: attributesPlugin)
         }
     }
 
     func registerDeviceToken(_ deviceToken: String) {
+        logger.debug("storing device token to device storage \(deviceToken)")
+        // save the device token for later use.
+        // segment plugin doesn't store token anywhere so we need to pass token to it every time
+        // storing it so we can reference the token and register on app relaunch
+        globalDataStore.pushDeviceToken = deviceToken
+        setDeviceToken(deviceToken)
+    }
+
+    /// Internal method for passing the device token to the plugin
+    private func setDeviceToken(_ deviceToken: String) {
+        logger.info("registering device token \(deviceToken)")
+        // Add device attributes first so they are updated for unsupported OS too
+        addDeviceAttributes()
+
+        // OS name might not be available if running on non-apple product. We currently only support iOS for the SDK
+        // and iOS should always be non-nil. Though, we are consolidating all Apple platforms under iOS but this check
+        // is required to prevent SDK execution for unsupported OS.
+        if deviceInfo.osName == nil {
+            logger.info("SDK being executed from unsupported OS. Ignoring request to register push token.")
+            return
+        }
         analytics.setDeviceToken(deviceToken)
```

Comment:
can we use the same plugin for token and attributes?

Reply:
Updated